### PR TITLE
Zone pivot validations

### PR DESCRIPTION
**Pull request by bianyifan**
[AB#323916](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/323916)

[AB#323917](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/323917)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6670)

### CODE REVIEW

File: src/docfx/build/metadata/UserMetadata.cs [hunk 1]
```diff
@@ -52,6 +52,10 @@ internal class UserMetadata
 
         public string? OriginalContentGitUrlTemplate { get; private set; }
 
+        public string? ZonePivotGroupFilename { get; private set; }
```
Comment (yufeih) [inline src/docfx/build/metadata/UserMetadata.cs:55]: Add a comment that this is the file name of the published URL with .json extension, not the source file name.

See an example [here](https://github.com/dotnet/AspNetCore.Docs/blob/93f237ef4ddf9f68ab29bd7a1ba6f9e2c6904a36/aspnetcore/docfx.json#L60)

[This](https://dev.azure.com/ceapex/Engineering/_git/docs-ui?path=%2Fpackages%2Fscripts%2Fsrc%2Fcomponents%2Fzone-pivot.ts&version=GBmaster&line=310&lineEnd=310&lineStartColumn=17&lineEndColumn=34&lineStyle=plain) is how docs-ui loads the zone pivot files.

Adjust the implementation accordingly to load the zone pivot definition file based on publish URL. You can get the publish URL from `PublishUrlMap.GetFilesByUrl` method.

File: src/docfx/build/zonepivot/ZonePivotProvider.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        private readonly LinkResolver _linkResolver;
+
+        private readonly ConcurrentDictionary<FilePath, FilePath?> _zonePivotDefinitionFileCache = new ConcurrentDictionary<FilePath, FilePath?>();
+        private readonly ConcurrentDictionary<FilePath, ZonePivotGroupDefinitionModel?> _zonePivotDefinitionModelCache =
+            new ConcurrentDictionary<FilePath, ZonePivotGroupDefinitionModel?>();
+
+        public ZonePivotProvider(
+            ErrorBuilder errors,
+            MetadataProvider metadataProvider,
+            FileResolver fileResolver,
```
Comment (yufeih) [inline src/docfx/build/zonepivot/ZonePivotProvider.cs:26]: `FileResolver` is used to resolve non-content files (typically configuration). Use `Input` to read content files.

File: src/docfx/Errors.cs [hunk 3]
```diff
@@ -479,6 +479,27 @@ public static Error MonikeRangeOutOfScope(SourceInfo<string?> configMonikerRange
                 => new Error(ErrorLevel.Error, "moniker-range-out-of-scope", $"No moniker intersection between docfx.yml/docfx.json and file metadata. Config moniker range '{configMonikerRange}' is {StringUtility.Join(configMonikers)}, while file monikers is {StringUtility.Join(fileMonikers)}.", source);
         }
 
+        public static class ZonePivot
+        {
+            public static Error ZonePivotGroupDefinitionNotFound(FilePath file)
```
Comment (yufeih) [inline src/docfx/Errors.cs:484]: This error does not appear in the feature spec, is it the same as `pivot-group-not-found`?

**Review by yufeih — CHANGES_REQUESTED** [review]

**Review by yufeih — COMMENTED** [review]
Two main comments:

1. The retrieval of zone pivot group file is based on publish url.
2. It is best practice to put code and test changes in a single PR.

File: src/docfx/Errors.cs (hunk 3 above)
Comment (bianyifan) [reply src/docfx/Errors.cs:484]: Yes, this is not in the feature spec, and it's for the case where `zone_pivot_groups` presents but we can't find `zone-pivot-groups.json` or `zone_pivot_groups_filename` in metadata.

Comment (yufeih) [reply src/docfx/Errors.cs:484]: For that we have a general `file-not-found` error.

File: src/docfx/build/zonepivot/ZonePivotProvider.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                return null;
+            }
+        }
+
+        private FilePath? GetZonePivotGroupDefinitionFile(FilePath file)
+        {
+            return _zonePivotDefinitionFileCache.GetOrAdd(file, _ =>
+                {
+                    var definitionFile = _metadataProvider.GetMetadata(_errors, file).ZonePivotGroupFilename ?? "zone-pivot-groups.json";
+                    var files = _publishUrlMap.Value.GetFilesByUrl("/" + definitionFile).ToList();
```
Comment (yufeih) [inline src/docfx/build/zonepivot/ZonePivotProvider.cs:95]: prepend `Config.BasePath`

File: src/docfx/build/zonepivot/ZonePivotProvider.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            Input input,
+            Lazy<PublishUrlMap> publishUrlMap)
+        {
+            _errors = errors;
+            _metadataProvider = metadataProvider;
+            _input = input;
+            _publishUrlMap = publishUrlMap;
+        }
+
+        public (FilePath?, ZonePivotGroup?) GetZonePivotGroup(FilePath file)
```
Comment (yufeih) [inline src/docfx/build/zonepivot/ZonePivotProvider.cs:35]: Change the return type to ensure that either there is no pivot group (`null`), or there is a valid one (both non-null). Internally they could share a single dictionary.
```suggestion
        public (FilePath, ZonePivotGroup)? TryGetZonePivotGroup(FilePath file)
```

File: src/docfx/validation/ContentValidator.cs [hunk 6]
```diff
@@ -215,6 +219,31 @@ public void ValidateTocEntryDuplicated(FilePath file, List<FilePath> referencedF
...
+
+            foreach (var (tripleColon, pivotId) in zonePivotUsages)
+            {
+                if (usage.ContainsKey(pivotId))
+                {
+                    usage[pivotId] = true;
+                }
+                else
+                {
+                    _errors.Add(Errors.ZonePivot.ZonePivotIdNotFound(tripleColon.GetSourceInfo(), pivotId, group!.Id, definitionFile));
```
Comment (yufeih) [inline src/docfx/validation/ContentValidator.cs:240]: Avoid using `!` to assert nullability whenever possible.

File: src/docfx/validation/ContentValidator.cs [hunk 7]
```diff
@@ -215,6 +219,31 @@ public void ValidateTocEntryDuplicated(FilePath file, List<FilePath> referencedF
             }
         }
 
+        public void ValidateZonePivots(FilePath file, List<(TripleColonBlock, string)> zonePivotUsages)
+        {
+            var (definitionFile, group) = _zonePivotProvider.GetZonePivotGroup(file);
+            var usage = group == null ? new Dictionary<string, bool>() : group.Pivots.Select(p => p.Id).Distinct().ToDictionary(p => p, _ => false);
+            if ((definitionFile == null || group == null) && zonePivotUsages.Any())
+            {
+                // TODO: throw error because we are unable to load definition file or group, yet got zone pivot usages
```
Comment (yufeih) [inline src/docfx/validation/ContentValidator.cs:228]: consider handle this error in `ZonePivotProvider.cs`.

File: src/docfx/validation/ContentValidator.cs [hunk 8]
```diff
@@ -215,6 +219,31 @@ public void ValidateTocEntryDuplicated(FilePath file, List<FilePath> referencedF
             }
         }
 
+        public void ValidateZonePivots(FilePath file, List<(TripleColonBlock, string)> zonePivotUsages)
```
Comment (yufeih) [inline src/docfx/validation/ContentValidator.cs:222]: nit: Take `SourceInfo<T>` as other methods like `ValidateImageLink`
```suggestion
        public void ValidateZonePivots(FilePath file, List<SourceInfo<string> zonePivotUsages)
```

File: src/docfx/build/zonepivot/ZonePivotProvider.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        /// Get publish URL of zone pivots definition file.
+        /// </summary>
+        /// <param name="definitionFilename">The definition filename from metadata, if null, use default "zone-pivot-groups.json".</param>
+        /// <returns>Published URL of zone pivots definition file.</returns>
+        private string GetZonePivotDefinitionPublishUrl(string? definitionFilename)
+        {
+            var filename = definitionFilename ?? s_defaultDefinitionFile;
+            return "/" + UrlUtility.Combine(
+                _config.BasePath.Value.Split('/').FirstOrDefault(),
+                filename.ReplaceRegex(s_extensionReplacementRegex, ".json"));
```
[the remark this answers is not in the document]
Comment (bianyifan) [reply src/docfx/build/zonepivot/ZonePivotProvider.cs:157]: docs-ui did this in `packages/docfx-templates/chrome.common.js` (L312)

[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/zonepivot/ZonePivotProvider.cs:157]: ok, then there is a handy `Path.ChangeExtension` method just for that purpose

**Review by yufeih — APPROVED** [review]

File: src/docfx/build/zonepivot/ZonePivotProvider.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        /// <summary>
+        /// Get publish URL of zone pivots definition file.
+        /// </summary>
+        /// <param name="file">Path of the file referencing zone pivot groups</param>
+        /// <param name="definitionFilename">The definition filename from metadata, if null, use default "zone-pivot-groups.json".</param>
+        /// <returns>Published URL of zone pivots definition file.</returns>
+        private string GetZonePivotDefinitionPublishUrl(FilePath file, string? definitionFilename)
+        {
+            return "/" + PathUtility.NormalizeFile(UrlUtility.Combine(
+                _publishUrlMap.Value.TryGetPublishUrl(file).Split('/', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault(),
```
Comment (yufeih) [inline src/docfx/build/zonepivot/ZonePivotProvider.cs:156]: There is an existing `DocumentProvider.GetSiteUrl` method to retrieve the url.

**Comment by bianyifan** [comment]
I have checked the regression test results, they are all content issue to be expected.

Hold this PR to wait for PM to confirm the error messages.

File: src/docfx/Errors.cs [hunk 11]
```diff
@@ -479,6 +479,36 @@ public static Error MonikeRangeOutOfScope(SourceInfo<string?> configMonikerRange
                 => new Error(ErrorLevel.Error, "moniker-range-out-of-scope", $"No moniker intersection between docfx.yml/docfx.json and file metadata. Config moniker range '{configMonikerRange}' is {StringUtility.Join(configMonikers)}, while file monikers is {StringUtility.Join(fileMonikers)}.", source);
         }
 
+        public static class ZonePivot
```
Comment (OsmondJiang) [inline src/docfx/Errors.cs:482]: let's check with Megan whether she need manage these errors. I saw some repos are impacted by these errors

**Comment by bianyifan** [comment]
Will open another PR to get a clean PR history.

**Pull request closed without merging**